packages/headless/src/api/generated-answer/generated-answer-client.ts

```diff
@@ -0,0 +1,152 @@
+import {EventSourcePolyfill} from 'event-source-polyfill';
+import {Logger} from 'pino';
+import {SearchAppState} from '../..';
+import {AsyncThunkOptions} from '../../app/async-thunk-options';
+import {ClientThunkExtraArguments} from '../../app/thunk-extra-arguments';
+import {SSEErrorPayload} from '../../features/generated-answer/generated-answer-actions';
+import {URLPath} from '../../utils/url-utils';
+import {resetTimeout} from '../../utils/utils';
+import {SearchAPIClient} from '../search/search-api-client';
+import {
+  GeneratedAnswerCitationsPayload,
+  GeneratedAnswerMessagePayload,
+  GeneratedAnswerPayloadType,
+  GeneratedAnswerStreamEventData,
+  GeneratedAnswerStreamFinishReason,
+} from './generated-answer-event-payload';
+import {GeneratedAnswerStreamRequest} from './generated-answer-request';
+
+export interface GeneratedAnswerAPIClientOptions {
+  logger: Logger;
+}
+
+export interface AsyncThunkGeneratedAnswerOptions<
+  T extends Partial<SearchAppState>
+> extends AsyncThunkOptions<
+    T,
+    ClientThunkExtraArguments<SearchAPIClient, GeneratedAnswerAPIClient>
+  > {}
+
+const buildStreamingUrl = (url: string, orgId: string, streamId: string) =>
+  new URLPath(
+    `${url}/rest/organizations/${orgId}/machinelearning/streaming/${streamId}`
+  );
+
+const MAX_RETRIES = 3;
+const MAX_TIMEOUT = 5000;
+export const RETRYABLE_STREAM_ERROR_CODE = 1;
+
+export class GeneratedAnswerAPIClient {
+  private logger: Logger;
+
+  constructor(private options: GeneratedAnswerAPIClientOptions) {
+    this.logger = options.logger;
+  }
+
+  streamGeneratedAnswer(
+    params: GeneratedAnswerStreamRequest,
+    onMessage: (payload: GeneratedAnswerMessagePayload) => void,
+    onCitations: (payload: GeneratedAnswerCitationsPayload) => void,
+    onError: (payload: SSEErrorPayload) => void,
+    onCompleted: () => void
+  ) {
+    const {url, organizationId, streamId, accessToken} = params;
+
+    if (!streamId) {
+      this.logger.error('No stream ID found');
+      return;
+    }
+
+    let retryCount = 0;
+    let timeout: ReturnType<typeof setTimeout>;
+    let source: EventSourcePolyfill;
+
+    const checkAndRetry = (e: unknown): EventSourcePolyfill | void => {
```

Comment:
I haven't gone into details of everything happening here, but I * think * all the retry logic could be replaced/improved with `exponential-backoffs` dependency which is already in used in the project.

Reply:
I'll try it out!

Reply:
Hmmm i tried replacing it but what i can't seem to replicate is that this mechanism is intended to detect when it's been too long between message events. The backOff function expects the promise function that can be rejected which doesn't happen with this SSE stream.